File: api-gateway/internal/handlers/data_loading_strategy_golden_test.go
package handlers

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

// TestDataLoadingStrategyStepsMatchSharedGolden pins dataLoadingStrategySteps to
// shared/data_loading_strategy_golden.json, which the frontend's
// computeStrategySteps is pinned to as well. The frontend rebuilds these lines
// when a strategy arrives without explanation_steps, so a change made to one
// builder and not the other shows the same pipeline two ways on two pages.
func TestDataLoadingStrategyStepsMatchSharedGolden(t *testing.T) {
	path := filepath.Join("..", "..", "..", "shared", "data_loading_strategy_golden.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read golden %s: %v", path, err)
	}
	var cases []struct {
		Name  string `json:"name"`
		Input struct {
			Mode             string `json:"mode"`
			EffectiveCDCMode string `json:"effective_cdc_mode"`
			ScheduleStatus   string `json:"schedule_status"`
			RerunDefault     string `json:"rerun_default"`
			Dataset          string `json:"dataset"`
		} `json:"input"`
		Steps []string `json:"steps"`
	}
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatalf("parse golden: %v", err)
	}
	if len(cases) == 0 {
		t.Fatal("golden fixture is empty")
	}
	for _, c := range cases {
		in := c.Input
		got := dataLoadingStrategySteps(in.Mode, in.EffectiveCDCMode, in.ScheduleStatus, in.RerunDefault, in.Dataset)
		if !reflect.DeepEqual(got, c.Steps) {
			t.Errorf("%s:\n got: %q\nwant: %q", c.Name, got, c.Steps)
		}
	}
}

File: frontend/src/lib/pipeline/__tests__/dataLoadingStrategyGolden.test.ts
import { readFileSync } from "node:fs"
import { resolve } from "node:path"
import { describe, it, expect } from "vitest"
import { computeStrategySteps, type DataLoadingStrategy } from "../dataLoadingStrategy"

/**
 * Pins the frontend's fallback step builder to the fixture the api-gateway's
 * dataLoadingStrategySteps is also pinned to (data_loading_strategy_golden_test.go).
 * The server sends these lines as explanation_steps; this builder writes them when
 * a strategy arrives without them (chat confirmation, older payloads), so a change
 * to one and not the other shows the same pipeline two ways.
 */
type GoldenCase = { name: string; input: DataLoadingStrategy; steps: string[] }

const goldenPath = resolve(__dirname, "../../../../../shared/data_loading_strategy_golden.json")
const golden: GoldenCase[] = JSON.parse(readFileSync(goldenPath, "utf8"))

describe("computeStrategySteps matches the server's shared golden", () => {
  it("golden fixture is not empty", () => {
    expect(golden.length).toBeGreaterThan(0)
  })

  it.each(golden.map((g) => [g.name, g.input, g.steps] as const))("%s", (_name, input, want) => {
    expect(computeStrategySteps(input)).toEqual(want)
  })

  it("server-sent steps win over the fallback", () => {
    const sent = ["from the server"]
    expect(computeStrategySteps({ ...golden[0].input, explanation_steps: sent })).toEqual(sent)
  })
})
